change encryption scheme (back) to random keys

sahib opened the pull request:
See #66

sahib committed:
change encryption scheme (back) to random keys

sahib committed:
trying to fix a flaky test

sahib committed:
gateway: fix timing issue in event test

sahib committed:
run "go get -u ./..." and fix errors

sahib committed:
travis: fix task binary location

evgmik commented on util/std.go:
```diff
@@ -494,6 +494,43 @@ func PeekHeader(r io.ReadSeeker, size int64) ([]byte, io.ReadSeeker, error) {
+
+func (hr *HeaderReader) Read(buf []byte) (int, error) {
+	n, err := hr.r.Read(buf)
+	if diff := hr.size - uint64(len(hr.buf)); n > 0 && diff > 0 {
```
Could this potentially backfire? `uint` minus another `uint`, might result in a negative number, which would appear humongous after casting to `uint`.  I remember fixing such problem in repiner.

sahib replied:
Oops, right. Luckily that case was caught by the inner if. Changed to a proper ``int64()`` here.

evgmik reviewed (commented):
Looks like we lost our deduplication. We might search for old key based on Content Hash and reuse key and back end hash from this.

Other than my comment about `util/std.go`, changes seems to be good.

 The rest look good to me

evgmik commented on catfs/fs.go:
```diff
@@ -1021,6 +985,16 @@ func (fs *FS) Stage(path string, r io.ReadSeeker) error {
+	// The stream was consumed, we now know those attrs:
+	size := sizeAcc.Size()
+	contentHash := hashWriter.Finalize()
+	compressAlgo, err = compress.GuessAlgorithm(path, headerReader.Header())
```
We should run guessing before fs.bk.Add, here it useless and potentially incorrect.
By this time we already assigned default `comressAlgo` and used it with `mio` in `fs.bk.Add` couple lines above.

sahib replied:
Right, I had a logic error here. Now I remember why ``PeekHeader()`` actually did the reading before the add. :joy:
Seems I need to use the existing ``prefixReader`` but patch the part I read to the stream. Luckily Go makes that easy, that would be quite a horror in C/C++. Too bad I wrote some nice tests for ``HeaderReader``.

sahib commented:
> Looks like we lost our deduplication. We might search for old key based on Content Hash and reuse key and back end hash from this.

Might be an idea, but another partial fix. We use [AES-GCM](https://de.wikipedia.org/wiki/Galois/Counter_Mode) as cipher mode for encryption. This means effectively: one changed bit in the beginning yields a completely different cipher text anyways and kills deduplication. This needs a more general fix.

In an ideal world IPFs would do the encryption for us and we would not bother for the deduplication since IPFS would handle that for us. Alternatively we would have to do our own chunking or do some like git's [packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) (i.e. re-encode several versions of a file to a highly compressed pack).

sahib commented:
By the way, you found two good points. :+1: Feel free to click "request changes" next time, not just comment. :smirk:

sahib committed:
review comment: fix types in HeaderReader

sahib committed:
tests: make badger shut up

sahib committed:
go.mod: downgrade fuse since it broke tests

sahib committed:
fs: read the header buffer before actually adding to ipfs for mime guessing (review comment)

sahib committed:
Merge branch 'develop' of github.com:sahib/brig into optimization/random-encryption-scheme

sahib committed:
go mod tidy

sahib committed:
run linter

sahib committed:
do not use io.MultiReader since it upsets compression/encryption Writer...

sahib commented:
Side note: It seems that [upgrading FUSE breaks it](https://github.com/sahib/brig/pull/69/commits/d9514f4823c55967eb0b60085fed3715dd623744). I didn't really check what the cause was, but I think we should update since we're still using a version from 2018. Probably best to incrementally update and see where (and actually what) breaks.

@evgmik: You want to have a look on that?

evgmik commented:
> > Looks like we lost our deduplication. We might search for old key based on Content Hash and reuse key and back end hash from this.
> 
> Might be an idea, but another partial fix. We use [AES-GCM](https://de.wikipedia.org/wiki/Galois/Counter_Mode) as cipher mode for encryption. This means effectively: one changed bit in the beginning yields a completely different cipher text anyways and kills deduplication. This needs a more general fix.
> 
> In an ideal world IPFs would do the encryption for us and we would not bother for the deduplication since IPFS would handle that for us. Alternatively we would have to do our own chunking or do some like git's [packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) (i.e. re-encode several versions of a file to a highly compressed pack).

I was thinking about a simplier case. What if we add files with different names but the same content (i.e. they are copies). In the old scheme, the key was derived from the content, so they end up with the same backend hash.  With random key, they endup with different backend hash but the same content hash. Since content is the same, we can check if we already have the same content hach, and use backend hash info from that node.

sahib commented:
> I was thinking about a simplier case. What if we add files with different names but the same content (i.e. they are copies). In the old scheme, the key was derived from the content, so they end up with the same backend hash. With random key, they endup with different backend hash but the same content hash. Since content is the same, we can check if we already have the same content hach, and use backend hash info from that node.

Yes, I understood that - that's why I called it a partial fix. It makes the situation a little better, but not much. Thus I don't consider it worthy to implement for now. The reason is that we use AES-GCM - changing a single bit early in the file will lead to a completely different text. Adding completely same content will result in deduplicated blocks (hopefully), but adding very similar files or modifying files will not.

Also there's a similar hen-egg problem as with the compression algorithm: to find out that we have the same content hash we need to read once through the stream (unless we do two passes as before). At that point the file is already encrypted.

sahib committed:
encrypt/compress: fix ReadFrom to always a sensible amount

sahib committed:
bug fix: compress: handle big read buffers correctly

sahib committed:
Merge branch 'develop' into optimization/random-encryption-scheme

evgmik commented on catfs/fs.go:
```diff
@@ -955,35 +955,36 @@ func (fs *FS) Stage(path string, r io.Reader) error {
+		headerReader := util.NewHeaderReader(r, 2048)
+		headerBuf, err := headerReader.Peek()
 		if err != nil {
-			log.WithError(err).Warnf("failed to guess suitable zip algo for %s", path)
```
Would failed attempt of `headerReader.Peek()` move the cursor position in `r`? Shall we reset it back to zero?

sahib replied:
Yes, it does move the cursor position, but headerReader preprends the buffer it read back to the stream.

evgmik commented on catfs/fs.go:
```diff
@@ -955,35 +955,36 @@ func (fs *FS) Stage(path string, r io.Reader) error {
 		fs.cfg.String("compress.default_algo"),
 	)
 
+	// We might not be able to read the header for some reason.
```
I would move the definition above the read of the default compression. Otherwise `err` check is visually disconnected from the code producing it.

sahib replied:
Done.

evgmik commented on fuse/file.go:
```diff
@@ -3,12 +3,10 @@
 	"os"
 	"sync"
-	"context"
-	"fmt"
```
looks like linter went crazy. We need `fmt` and `log` for this module to compile

sahib replied:
No, I went crazy. Will fix once I found that bug.

sahib commented:
> If guess failed, `compressAlgo` potentially could be not set or damaged. We should add to the error handler:
> 
> ```
> compressAlgo = compress.AlgoNone
> ```
> 
> Since we use it in the couple places. We need to define `compressFallBack = compress.AlgoNone`

The zero value of ``compressAlgo`` is AlgoNone. So also when the variable would not be set, it would default to that.

sahib committed:
Temporary commit during bug hunt (also updated go-ipfs-api)

I added some utilities that help looking at this strange bug:

* brig debug ds: decode a stream
* brig debug es: encode a stream
* brig debug isb: Simulate the bug by using our internal `ipfs cat`.

Here's one way to trigger the bug:

$ dd if=/dev/urandom of=$HOME/256M bs=1M count=256
$ cat ~/256M | brig debug es --key GEoPSbu31Dpypn9Tmy5GsEaiM8iKhNcohy5W7nasjzj4 | ipfs add -
$ brig debug isb --key GEoPSbu31Dpypn9Tmy5GsEaiM8iKhNcohy5W7nasjzj4 --hash QmNNjFN4pTpUwY7AesJhpEJ1kENBbYyXaeBmYcCBLu7MpC  > /tmp/x; du -hs /tmp/x

You might have to execute the last line several times to trigger the
issue.  Sometimes it will show 256M, sometimes just 37M or any number.
I will try to further minimize the test case.

Additionally fixed some review comments.

sahib committed:
Merge branch 'optimization/random-encryption-scheme' of github.com:sahib/brig into optimization/random-encryption-scheme

sahib committed:
fix messy merge and run linter

sahib committed:
bugfix: encrypt: always read full nonce, even if underlying reader gets less

This fixes the issue with the nonce size reported wrong and thus
erroring out early in the stream. The reason was that io.Reader.Read()
never guarantees that the passed buffer gets filled completely. I knew
this, but apparently not while writing the encrypt code.

This bug was there since a long time, but very seldomly showed. Only
with the last few IPFS releases it happened more often since they might
do more aggressive buffering and those might have short phases where
some bytes where not available yet and Read() returned early with less
bytes than we expected. Fixed by using io.ReadFull().

As a learning I will add a few commits with the following changes:

* Add a regression test (bash-based) that adds bigger files.
* Make the mock backend use very small buffers and/or do not always
  fill them completely. This should make sure this bug class will not
  happen again by simulating IPFS better.

sahib committed:
small cleanup of bug-hunt traces and scripts/ dir

sahib committed:
cmd: debug: add ten-{source,sink} as test data generator

This is just a useful utility that helped during the last bug hunt.
It can be used like this:

$ brig debug ten-source | brig debug ten-sink

If you have any idempotent stream operation you can use it like this:

$ brig debug ten-source --size 100M | \
    brig debug encode-stream | \
    brig debug decode-stream | \
    brig debug ten-sink --size 100M

You can of course also throw some `ipfs add` or `ipfs cat` in there.
Extend this whenever you need to debug weird streaming bugs.

In the future I want to add more such debug utilities to make brig
easier to explore for developers and users.

sahib committed:
catfs: mem backend's Cat uses (pseudo) random read buf sizes now

This should be good for tests where we want to detect if a writer
misbehaves.

evgmik commented:
Why the test commands are called ten-....?  I do not see logic of this name.

sahib commented:
> Why the test commands are called ten-....? I do not see logic of this name.

``ten`` because it prints (or validates) the numbers 0-9 (i.e. first ten). Liked it because it's short. Other suggestions?



Side note: I actually broke a PR best practice, because I got a little distracted during bug hunt. That PR got a little big and does more than what was initially advertised, should have done several smaller ones.

sahib committed:
merge with develop

evgmik commented:
Regarding "ten". I understand now why it called that way. I see no other way to call it concisely.
